src/topology/instances/add_circle.lean

```diff
@@ -286,3 +286,115 @@ local attribute [instance] fact_zero_lt_one
 /-- The unit circle `ℝ ⧸ ℤ`. -/
 @[derive [compact_space, normal_space, second_countable_topology]]
 abbreviation unit_add_circle := add_circle (1 : ℝ)
+
+
+section identify_Icc_ends
+/-! This section proves that the natural map from `[0, T] ⊂ ℝ` to `add_circle T` gives an
+identification of `add_circle T`, as a topological space, with the quotient of `[0, T]` by the
+equivalence relation identifying `0` with `T`. -/
+
+variables (T : ℝ) [hT : fact (0 < T)]
+include hT
+local notation `𝕋` := add_circle T
+
+private lemma Icc_to_circle_kernel {x y : ℝ} (hx : x ∈ Icc 0 T) (hy : y ∈ Icc 0 T) :
+  (x : 𝕋) = (y : 𝕋) ↔ (x = y) ∨ (x = 0 ∧ y = T) ∨ (y = 0 ∧ x = T) :=
+begin
+  simp_rw [quotient_add_group.eq, add_subgroup.mem_zmultiples_iff, zsmul_eq_mul],
+  split,
+  { rintros ⟨n, hn⟩,
+    have : | -x + y| ≤ T,
+    { rw abs_le,
+      exact ⟨by linarith [hx.2, hy.1], by linarith [hx.1, hy.2]⟩, },
+    rw [←hn, abs_mul, abs_of_pos hT.out, mul_le_iff_le_one_left hT.out, ←int.cast_abs,
+      ←int.cast_one, int.cast_le] at this,
+    replace this : |n| = 0 ∨ |n| = 1,
+    { rcases eq_or_lt_of_le this with h|h,
+      { exact or.inr h, },
+      { exact or.inl (by linarith [abs_nonneg n]) } },
+    rw [abs_eq (zero_le_one' ℤ), abs_eq_zero] at this,
+    rcases this with (rfl | rfl | rfl),
+    { rw int.cast_zero at hn,
+      exact or.inl (by linarith), },
+    { rw int.cast_one at hn,
+      exact or.inr (or.inl ⟨by linarith [hx.1, hy.2], by linarith [hx.1, hy.2]⟩) },
+    { rw [int.cast_neg, int.cast_one] at hn,
+      exact or.inr (or.inr ⟨by linarith [hy.1, hx.2], by linarith [hy.1, hx.2]⟩), }, },
+  { rintros (rfl | ⟨rfl, h⟩ | ⟨rfl, h⟩),
+    exacts [ ⟨0, by simp⟩, ⟨1, by {rw h, simp}⟩, ⟨-1, by {rw h, simp}⟩] },
+end
```

Comment:
It's probably a good idea to break this proof up into some smaller pieces. Here's one way to do that:
```
lemma coe_eq_zero_iff_of_mem_Ico {x : ℝ} (hx : x ∈ Ico 0 T) : (x : 𝕋) = 0 ↔ x = 0 :=
sorry

lemma coe_eq_coe_iff_of_mem_Ico {x y : ℝ} (hx : x ∈ Ico 0 T) (hx : y ∈ Ico 0 T) :
  (x : 𝕋) = y ↔ x = y :=
sorry

@[simp] lemma coe_period : (T : 𝕋) = 0 :=
by simp only [quotient_add_group.eq_zero_iff, add_subgroup.mem_zmultiples]

private lemma Icc_to_circle_kernel {x y : ℝ} (hx : x ∈ Icc 0 T) (hy : y ∈ Icc 0 T) :
  (x : 𝕋) = (y : 𝕋) ↔ (x = y) ∨ (x = 0 ∧ y = T) ∨ (y = 0 ∧ x = T) :=
begin
  obtain ⟨rfl | hx', rfl | hy'⟩ := ⟨eq_or_ne T x, eq_or_ne T y⟩;
    try { replace hx : x ∈ Ico 0 T := ⟨hx.1, lt_of_le_of_ne' hx.2 hx'⟩ };
    try { replace hy : y ∈ Ico 0 T := ⟨hy.1, lt_of_le_of_ne' hy.2 hy'⟩ },
  { simp },
  { simp only [coe_period, and_true, false_or, eq_self_iff_true, and_false, hy', hy'.symm,
      @eq_comm _ _ ↑y, coe_eq_zero_iff_of_mem_Ico _ hy], },
  { simp only [coe_period, and_true, false_or, eq_self_iff_true, or_false, and_false, hx',
      hx'.symm, coe_eq_zero_iff_of_mem_Ico _ hx], },
  { simp [or_false, and_false, or_self, hx'.symm, hy'.symm,
      coe_eq_coe_iff_of_mem_Ico _ hx hy]  },
end
```